src/domain/auth-rebac/domain/src/repos/rebac_repository.rs

```diff
@@ -0,0 +1,378 @@
+        object_entity: &Entity,
+    ) -> Result<(), DeleteEntitiesRelationError>;
+
+    async fn get_subject_entity_relations(
+        &self,
+        subject_entity: &Entity,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsError>;
+
+    async fn get_subject_entity_relations_by_object_type(
+        &self,
+        subject_entity: &Entity,
+        object_entity_type: EntityType,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsByObjectTypeError>;
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetRelationsBetweenEntitiesError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// DTOs
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum EntityType {
+    Dataset,
+    Account,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Entity<'a> {
+    pub entity_type: EntityType,
+    pub entity_id: Cow<'a, str>,
+}
+
+impl<'a> Entity<'a> {
+    pub fn new(entity_type: EntityType, entity_id: impl Into<Cow<'a, str>>) -> Self {
+        Self {
+            entity_type,
+            entity_id: entity_id.into(),
+        }
+    }
+
+    pub fn new_account(entity_id: impl Into<Cow<'a, str>>) -> Self {
+        Self::new(EntityType::Account, entity_id)
+    }
+
+    pub fn new_dataset(entity_id: impl Into<Cow<'a, str>>) -> Self {
+        Self::new(EntityType::Dataset, entity_id)
+    }
+
+    pub fn into_owned(self) -> Entity<'static> {
+        Entity {
+            entity_type: self.entity_type,
+            entity_id: Cow::Owned(self.entity_id.into_owned()),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct EntityWithRelation<'a> {
+    pub entity: Entity<'a>,
+    pub relation: Relation,
+}
+
+impl<'a> EntityWithRelation<'a> {
+    pub fn new(entity: Entity<'a>, relation: Relation) -> Self {
+        Self { entity, relation }
+    }
+
+    pub fn new_account(entity_id: impl Into<Cow<'a, str>>, relation: Relation) -> Self {
+        let account_entity = Entity::new_account(entity_id);
+
+        Self::new(account_entity, relation)
+    }
+
+    pub fn new_dataset(entity_id: impl Into<Cow<'a, str>>, relation: Relation) -> Self {
+        let dataset_entity = Entity::new_dataset(entity_id);
+
+        Self::new(dataset_entity, relation)
+    }
+}
+
+pub type ObjectEntity<'a> = Entity<'a>;
+pub type ObjectEntityWithRelation<'a> = EntityWithRelation<'a>;
```

Comment:
What's the value of these typedefs?

Reply:
Since these type-aliases are causing a second question, I decided to remove them

Removed